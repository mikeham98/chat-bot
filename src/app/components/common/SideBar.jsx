import React from 'react';
import PropTypes from 'prop-types';
import styles from '../../../themes/components/sideBar.scss';

const maxWidth = '300px';
const minWidth = '1px';

export default class SideBar extends React.PureComponent {
    constructor(props) {
        super(props);
        this.minSideBarWidth = minWidth;
        this.addTransition = this.addTransition.bind(this);
        this.removeTransition = this.removeTransition.bind(this);
        this.handleResize = this.handleResize.bind(this);
        this.handleBounce = this.handleBounce.bind(this);
        this.onMouseUp = this.onMouseUp.bind(this);
        this.onMouseDownDraggableHandle = this.onMouseDownDraggableHandle.bind(this);
    }

    componentDidMount() {
        // Add a mouse down listener to the handle so that the menu can be resized on mouse move
        this.sideBarDraggableHandle.addEventListener("mousedown", this.onMouseDownDraggableHandle);

        // On mouse up (any where on the document) bounce the size bar and remove the mouse down event handler
        document.addEventListener("mouseup", this.onMouseUp);
    }

    componentWillUnmount() {
        // clean up event listeners
        this.sideBarDraggableHandle.removeEventListener("mousedown", this.onMouseDownDraggableHandle);
        document.removeEventListener("mouseup", this.onMouseUp);
        document.removeEventListener("mousemove", this.handleResize);
        document.removeEventListener('selectstart', this.disableSelect);
    }

    onMouseUp() {
        this.handleBounce();
        document.removeEventListener("mousemove", this.handleResize);
        document.removeEventListener('selectstart', this.disableSelect);
    }

    onMouseDownDraggableHandle() {
        this.minSideBarWidth = this.sideBarWrapper.style.width;
        this.removeTransition();
        document.addEventListener("mousemove", this.handleResize);
        document.addEventListener('selectstart', this.disableSelect);
    }

    handleResize(event) {
        this.sideBarWrapper.style.width = `${event.x}px`;
    }

    disableSelect(event) {
        event.preventDefault();
    }

    handleBounce() {
        const sideBarWidth = this.sideBarWrapper.style.width;
        if (sideBarWidth !== maxWidth) {
            this.addTransition();
            if (sideBarWidth < maxWidth && sideBarWidth !== minWidth) {
                this.sideBarWrapper.style.width = minWidth;
                return;
            }
            this.sideBarWrapper.style.width = maxWidth;
        }
    }

    addTransition() {
        this.sideBarWrapper.style.transition = 'width 300ms';
    }

    removeTransition() {
        this.sideBarWrapper.style.transition = 'none';
    }

    render() {
        return (
            <div className={styles.sideBarContainer}>
                <div ref={ref => this.sideBarWrapper = ref} style={{width: 300}}>
                    <div className={styles.sideBarMenuWrapper}>
                        <div className={styles.sideBarMenu}>
                            <div style={{
                                width: '100%',
                                display: 'flex',
                                flexDirection: 'column'
                            }}>
                                <h1 className={styles.sideBarMenuTitle}>
                                    {this.props.title}
                                </h1>
                                <div className={styles.sideBarMenuChildren}>
                                    {this.props.children}
                                </div>
                            </div>
                        </div>
                        <div ref={(ref) => this.sideBarDraggableHandle = ref} className={styles.sideBarDraggableHandle}>
                        </div>
                    </div>
                </div>
            </div>
        )
    }
}

SideBar.propTypes = {
    title: PropTypes.string.isRequired,
    children: PropTypes.any,
};