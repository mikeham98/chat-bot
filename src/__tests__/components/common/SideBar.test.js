import React from 'react';
import {mount} from 'enzyme';
import SideBar from '../../../app/components/common/SideBar'

// This component would take a very long time to test therefore i have just done a simple snapshot

describe('SideBar', () => {
    let mountedWrapper;
    let props = {};
    const wrapper = () => {
        if (!mountedWrapper) {
            mountedWrapper = mount(<SideBar {...props}/>);
        }
        return mountedWrapper;
    };

    beforeEach(() => {
        props = {
            title: 'This is a side bar',
            children: <div>Children</div>
        };
        mountedWrapper = undefined;
    });

    describe('render', () => {
        it('should return a title of This is a side bar and a div with the text Children inside', () => {
            expect(wrapper()).toMatchSnapshot();
        });
    });
});