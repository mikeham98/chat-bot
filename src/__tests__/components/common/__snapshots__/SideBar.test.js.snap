// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SideBar render should return a title of This is a side bar and a div with the text Children inside 1`] = `
<SideBar
  title="This is a side bar"
>
  <div
    className="sideBarContainer"
  >
    <div
      id="sideBarWrapper"
      style={
        Object {
          "width": 300,
        }
      }
    >
      <div
        className="sideBarMenuWrapper"
        id="sideBarMenuWrapper"
      >
        <div
          className="sideBarMenu"
        >
          <div
            className="sideBarContent"
          >
            <div
              className="sideBarMenuTitle"
            >
              <h1>
                This is a side bar
              </h1>
            </div>
            <div
              className="sideBarMenuChildren"
            >
              <div>
                Children
              </div>
            </div>
          </div>
        </div>
        <div
          className="sideBarDraggableHandle"
        />
      </div>
    </div>
  </div>
</SideBar>
`;
